### PR TITLE
Proxy Client Metrics

### PR DESCRIPTION
**Pull request by travisgroth**
Proxy client metrics implementation for #35 

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] updated CHANGELOG.md
- [x] ready for review

### CODE REVIEW

File: internal/tripper/chain_test.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+			body, _ := ioutil.ReadAll(resp.Body)
+			mockResp := httptest.NewRecorder()
+			mockResp.Write(body)
+			mockResp.WriteString(fmt.Sprintf(",%s", id))
+			return mockResp.Result(), nil
+		})
+	}
+}
+
+func TestNew(t *testing.T) {
```
Comment (travisgroth) [inline internal/tripper/chain_test.go]: The tripper chain functionality is largely copied from middleware, but I wanted to redo the tests from first principles so that (a) I understood the code well (b) could focus on the side effects rather than returned data structures.  

I think this approach captured all the tests nicely, but I'm happy to port over the middleware chain tests if you'd like to keep it consistent.

File: internal/metrics/middleware.go [hunk 2]
```diff
@@ -79,10 +116,37 @@ func HTTPMetricsHandler(service string) func(next http.Handler) http.Handler {
...
+}
+
+// HTTPMetricsRoundTripper creates a metrics tracking tripper for outbound HTTP Requests
+func HTTPMetricsRoundTripper(service string) func(next http.RoundTripper) http.RoundTripper {
+
+	return func(next http.RoundTripper) http.RoundTripper {
+		return tripper.RoundTripperFunc(func(r *http.Request) (*http.Response, error) {
+			startTime := time.Now()
+
+			resp, err := next.RoundTrip(r)
```
Comment (desimone) [inline internal/metrics/middleware.go]: If an error gets thrown here, I assume `resp` would be nil and the next two lines may panic due to nil reference?

File: internal/tripper/chain_test.go (hunk 1 above)
Comment (desimone) [reply internal/tripper/chain_test.go]: I think that makes sense. I'm generally for a bit of duplication for the sake of clarity and I think that's true here.

Comment (travisgroth) [reply internal/tripper/chain_test.go]: Do you want me to bring over the middleware tests or leave the tests as-is?  I'm okay either way.

File: internal/metrics/middleware.go (hunk 2 above)
Comment (travisgroth) [reply internal/metrics/middleware.go]: That's a good point.  I was thinking on the server side still...and that a response of some sort would happen - but that's not true here in the Transport layer.  Will fix.

File: internal/tripper/chain_test.go (hunk 1 above)
Comment (travisgroth) [reply internal/tripper/chain_test.go]: Also an option: update the middleware tests to use this style.

Comment (desimone) [reply internal/tripper/chain_test.go]: Hmmm... If you think the middleware tests are relevant -- sure. If not, I don't think it's required. I'll defer to your judgement. 

The one thing that would be nice to add (sorry, I can't see coverage directly here as I'm traveling, so perhaps you already have that base covered) would be to have a mocked RoundTripper that also could test for that nil condition from before. .... maybe add a `roundTripError` type to `mockTransport`?

As for changing the existing middleware tests format, I'm a bit more cautious just because they are a complete carry over from [Alice](https://github.com/pomerium/pomerium/blob/master/3RD-PARTY#L90_ and have been pretty stable. Though, adding additional relevant tests makes sense.

Comment (travisgroth) [reply internal/tripper/chain_test.go]: >have a mocked RoundTripper that also could test for that nil condition from before

Yep, should be covered.  I have a then() transport/tripper that errors. 
 https://github.com/travisgroth/pomerium/blob/ca46202d21561eab90c5a59113080a5c5190b6c3/internal/metrics/middleware_test.go#L128-L132  and https://github.com/travisgroth/pomerium/blob/ca46202d21561eab90c5a59113080a5c5190b6c3/internal/metrics/middleware_test.go#L215-L221

>As for changing the existing middleware tests format, I'm a bit more cautious just because they are a complete carry over from [Alice]

Ah, I did not realize.  I think they accomplish the same thing; just a different approach.  Take a look when you're back at a keyboard for a bit.  I'm happy to leave as is; we can always revisit.

Comment (desimone) [reply internal/tripper/chain_test.go]: Ah! Figured that I just missed it.

Comment (desimone) [reply internal/tripper/chain_test.go]: > I think they accomplish the same thing; just a different approach. Take a look when you're back at a keyboard for a bit. I'm happy to leave as is; we can always revisit.

Yeah, let's revisit in another issue.

**Comment by desimone** [comment]
Thanks!!
